### PR TITLE
mcuboot: Add generating Zigbee OTA package

### PR DESCRIPTION
**Pull request by mariuszpos**
Add generating Zigbee OTA package

Signed-off-by: Mariusz Poslinski <mariusz.poslinski@nordicsemi.no>

### CODE REVIEW

File: scripts/bootloader/zb_ota/package.py [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+                                                os.path.getsize(file_name),
+                                                self.calculate_crc(32, file_name) & 0xFFFFFFFF,
+                                                bytes(open(file_name, 'rb').read()),
+                                                self.manufacturer_id,
+                                                self.image_type,
+                                                self.comment,
+                                                self.zigbee_ota_min_hw_version,
+                                                self.zigbee_ota_max_hw_version)
+
+                ota_file_handle = open(self.zigbee_ota_file.filename, 'wb')
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota/package.py:302]: can't `with` block statement be used here?

File: scripts/bootloader/zb_ota/package.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        files = os.listdir(work_dir)
+
+        with ZipFile(filename, 'w') as package:
+            for _file in files:
+                file_path = os.path.join(work_dir, _file)
+                package.write(file_path, _file)
+
+    @staticmethod
+    def calculate_file_size(firmware_filename):
+        b = os.path.getsize(firmware_filename)
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota/package.py:333]: `b` is unneeded

File: scripts/bootloader/zb_ota/package.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+        # return hash in little endian
+        sha256 = digest.digest()
+        return sha256[31::-1]
+
+    @staticmethod
+    def calculate_crc(crc, firmware_filename):
+        """
+        Calculates CRC16 has on provided firmware filename
+
+        :type str firmware_filename:
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota/package.py:360]: incomplete doc string

File: scripts/bootloader/zb_ota/package.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+            FirmwareKeys.INIT_PACKET_DATA: init_packet_data.copy()
+            }
+
+        if firmware_version is not None:
+            self.firmwares_data[firmware_type][FirmwareKeys.INIT_PACKET_DATA][PacketField.FW_VERSION] = firmware_version
+
+    @staticmethod
+    def normalize_firmware_to_bin(work_dir, firmware_path):
+        firmware_filename = os.path.basename(firmware_path)
+        new_filename = firmware_filename.replace(".hex", ".bin")
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota/package.py:397]: In theory there can be `.hex` strings inside filename (not in the end of a file) and these occurrences will be also replaced

File: scripts/bootloader/zb_ota_sign.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+    if zigbee:
+        from shutil import copyfile
+        from os import remove
+
+        log_message = "Zigbee update created at {0}".format(package.zigbee_ota_file.filename)
+        click.echo(log_message)
+
+        # Taking the inner Zigbee package as input for the outer DFU package
+        binfile = package.zigbee_ota_file.filename.replace(".zigbee", ".bin")
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota_sign.py:239]: In theory there can be `.zigbee` string inside filename (not in the end of a file) and these occurrences will be also replaced

File: scripts/bootloader/zb_ota/package.py [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+                            from_bytes = None,
+                            hash_bytes=firmware_hash,
+                            hash_type=HashTypes.SHA256,
+                            dfu_type=HexTypeToInitPacketFwTypemap[key],
+                            fw_version=firmware_data[FirmwareKeys.INIT_PACKET_DATA][PacketField.FW_VERSION],
+                            hw_version=firmware_data[FirmwareKeys.INIT_PACKET_DATA][PacketField.HW_VERSION],
+                            app_size=app_size)
+
+            # Store the .dat file in the work directory
+            init_packet_filename = firmware_data[FirmwareKeys.BIN_FILENAME].replace(".bin", ".dat")
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota/package.py:277]: In theory there can be `.bin` string inside filename (not in the end of a file) and these occurrences will be also replaced

File: scripts/bootloader/zb_ota/package.py [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+            raise PackageException(
+                "Target directory {0} exists, not able to unpack to that directory.",
+                target_dir)
+
+        with ZipFile(package_path, 'r') as pkg:
+            pkg.extractall(target_dir)
+
+            with open(os.path.join(target_dir, Package.MANIFEST_FILENAME), 'r') as f:
+                _json = f.read()
+                """:type :str """
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota/package.py:439]: is this line needed?

File: scripts/bootloader/zb_ota_sign.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+
+"""nrfutil command line tool."""
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota_sign.py:7]: not nrfutil anymore

File: scripts/bootloader/zb_ota_sign.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+def int_as_text_to_int(value):
+    try:
+        if value[:2].lower() == '0x':
+            return int(value[2:], 16)
+        elif value[:1] == '0':
+            return int(value, 8)
+        return int(value, 10)
+    except ValueError:
+        raise print('%s is not a valid integer' % value)
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota_sign.py:23]: invalid expression. print is not exception

File: scripts/bootloader/zb_ota_sign.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        try:
+            if value.lower() == 'none':
+                return 'none'
+            return int_as_text_to_int(value)
+        except ValueError:
+            print('%s is not a valid integer' % value, param, ctx)
+
+BASED_INT_OR_NONE = BasedIntOrNoneParamType()
+
+class BasedIntParamType(BasedIntOrNoneParamType):
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota_sign.py:38]: what's the difference between `BasedIntOrNoneParamType` and `BasedIntParamType`?

File: scripts/bootloader/zb_ota_sign.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+BASED_INT = BasedIntParamType()
+
+def convert_version_string_to_int(s):
+    """Convert from semver string "1.2.3", to integer 10203"""
+    numbers = s.split(".")
+    if len(numbers) != 3:
+        raise click.BadParameter("Must be on the format x.y.z", param_hint='application-version-string')
+    js = [10000, 100, 1]
+    return sum([js[i] * int(numbers[i]) for i in range(3)])
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_ota_sign.py:49]: Returned value can be same for different versions (for example 1.1.0 and 1.0.100). Is this method good enough? The max value of part of semver is 1024

File: scripts/bootloader/zb_ota/manifest.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        :type dict firmwares_data: The firmwares data structure describing the Nordic DFU package
+        """
+        self.firmwares_data = firmwares_data
+        self.manifest = None
+
+    def generate_manifest(self):
+        self.manifest = Manifest()
+
+        for key in self.firmwares_data:
+            firmware_dict = self.firmwares_data[key]
```
[the remark this answers is not in the document]
Comment (mariuszpos) [reply scripts/bootloader/zb_ota/manifest.py]: Done

File: cmake/mcuboot.cmake [hunk 13]
```diff
@@ -111,6 +115,24 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
...
+    set(zb_sign_cmd
+      ${PYTHON_EXECUTABLE}
+      ${NRF_DIR}/scripts/bootloader/zb_ota_sign.py
+      generate
+      --application ${PROJECT_BINARY_DIR}/app_signed.hex
+      --application-version-string ${CONFIG_ZIGBEE_APP_VERSION}
+      --hw-version ${CONFIG_ZIGBEE_HARDWARE_VERSION}
+      --zigbee-ota-hw-version ${CONFIG_ZIGBEE_OTA_HW_VERSION}
+      --zigbee-ota-fw-version ${CONFIG_ZIGBEE_OTA_FW_VERSION}
+      --zigbee-ota-min-hw-version ${CONFIG_ZIGBEE_OTA_MIN_HW_VERSION}
```
Comment (sebastiandraus) [inline cmake/mcuboot.cmake:128]: @mariuszpos Will it be possible to have `--zigbee-ota-min-hw-version` and `--zigbee-ota-max-hw-version` as options, so these attributes may not present in the Zigbee OTA header?

**Comment by sebastiandraus** [comment]
@mariuszpos What is `scripts/bootloader/app_signed.hex` used for?

File: scripts/bootloader/zb_ota/package.py [hunk 14]
```diff
@@ -0,0 +1,446 @@
...
+        binascii.hexlify(cmd.init.hash.hash),
+        )
+
+        return s
+
+    def __str__(self):
+
+        imgs = ""
+        i = 0
+        if self.manifest.application:
```
[the remark this answers is not in the document]
Comment (mariuszpos) [reply scripts/bootloader/zb_ota/package.py:216]: Currently it will be only 1 image. In future could be more, e.g. image for bootloader.

File: cmake/mcuboot.cmake (hunk 13 above)
Comment (sebastiandraus) [reply cmake/mcuboot.cmake:128]: Rationale: For jenkins tests we need to have it as options, so different sets of paramaters can be tested (no min/max hw version, min hw-version only etx.)

File: scripts/bootloader/zb_ota/package.py (hunk 4 above)
Comment (mariuszpos) [reply scripts/bootloader/zb_ota/package.py:386]: fixed

File: scripts/bootloader/zb_ota_sign.py (hunk 5 above)
Comment (mariuszpos) [reply scripts/bootloader/zb_ota_sign.py:238]: fixed

File: scripts/bootloader/zb_ota/package.py (hunk 6 above)
Comment (mariuszpos) [reply scripts/bootloader/zb_ota/package.py:268]: fixed

File: cmake/mcuboot.cmake [hunk 15]
```diff
@@ -111,6 +115,24 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
     --pad-header
     )
 
+  if(CONFIG_ZIGBEE)
+    set(zb_sign_cmd
+      ${PYTHON_EXECUTABLE}
+      ${NRF_DIR}/scripts/bootloader/zb_ota_sign.py
+      generate
+      --application ${PROJECT_BINARY_DIR}/app_signed.hex
```
Comment (sebastiandraus) [inline cmake/mcuboot.cmake:123]: @mariuszpos What about other fields of OTA header format, for example `image_type`? Could it be provided somewhere else?

File: scripts/bootloader/zb_ota/package.py [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+            "manifest": {
+                "bootloader": {
+                    "bin_file": "asdf.bin",
+                    "dat_file": "asdf.dat",
+                    "init_packet_data": {
+                        "application_version": null,
+                        "device_revision": null,
+                        "device_type": 5,
+                        "firmware_hash": "asdfasdkfjhasdkfjashfkjasfhaskjfhkjsdfhasjkhf",
+                        "softdevice_req": [
```
Comment (sebastiandraus) [inline scripts/bootloader/zb_ota/package.py:59]: Softdevice dependencies?

File: scripts/bootloader/zb_ota/dfu_cc_pb2.py [hunk 17]
```diff
@@ -0,0 +1,599 @@
...
+  full_name='dfu.FwType',
+  filename=None,
+  file=DESCRIPTOR,
+  values=[
+    _descriptor.EnumValueDescriptor(
+      name='APPLICATION', index=0, number=0,
+      serialized_options=None,
+      type=None),
+    _descriptor.EnumValueDescriptor(
+      name='SOFTDEVICE', index=1, number=1,
```
Comment (sebastiandraus) [inline scripts/bootloader/zb_ota/dfu_cc_pb2.py:60]: Softdevice dependencies? Is this file needed?

File: scripts/bootloader/zb_ota_sign.py [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+        click.echo('Warning: Non-ASCII characters in the comment are not allowed. Discarding comment.')
+        zigbee_comment = ''
+    elif len(zigbee_comment) > 30:
+        click.echo('Warning: truncating the comment to 30 bytes.')
+        zigbee_comment = zigbee_comment[:30]
+
+    if zigbee_manufacturer_id is None:
+        zigbee_manufacturer_id = 0xFFFF
+
+    if zigbee_image_type is None:
```
Comment (sebastiandraus) [inline scripts/bootloader/zb_ota_sign.py:186]: What about introducing some Enum to define very basic img_types, instead of using wildcards? For example, `zigbee_app` can be defined at first.

File: scripts/bootloader/app_signed.hex [hunk 19]
```diff
@@ -0,0 +1,13794 @@
+:020000040000FA
```
Comment (sigvartmh) [inline scripts/bootloader/app_signed.hex:1]: I assume this file is not intentionally included?

Comment (mariuszpos) [reply scripts/bootloader/app_signed.hex:1]: Yes, I'll remove it.

File: scripts/bootloader/zb_ota/dfu_cc_pb2.py [hunk 20]
```diff
@@ -0,0 +1,599 @@
+# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (wbober) [inline scripts/bootloader/zb_ota/dfu_cc_pb2.py:1]: This file should not be here. Protocol buffers are only used in nRF5 SDK.

File: scripts/bootloader/zb_ota/init_packet_pb.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
```
Comment (wbober) [inline scripts/bootloader/zb_ota/init_packet_pb.py:1]: This file should not be here. Protocol buffers are only used in nRF5 SDK.

File: scripts/bootloader/zb_ota/manifest.py [hunk 22]
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
```
Comment (wbober) [inline scripts/bootloader/zb_ota/manifest.py:1]: This file should not be here.

File: scripts/bootloader/zb_ota/model.py [hunk 23]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python3
```
Comment (wbober) [inline scripts/bootloader/zb_ota/model.py:1]: This file should not be here.

File: scripts/bootloader/zb_ota/ota_file.py [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+OTA_UPGRADE_FILE_HEADER_STACK_PRO     = 0x0002
+
+OTA_UPGRADE_SUBELEMENT_HEADER_SIZE    = 2 + 4  # Tag ID + Length Field
+OTA_UPGRADE_SUBELEMENT_TRIGGER_TYPE   = 0xCDEF
+OTA_UPGRADE_SUBELEMENT_TRIGGER_LENGTH = 1 + 4 + 4 + 4 + 4 # See background_dfu_trigger_t in the nRF5 SDK for explanation
+
+OTA_UPGRADE_FIELD_CONTROL_BIT_MASK_HW_VER = (1 << 2)
+OTA_UPGRADE_MIN_HW_VERSION_LENGTH         = 2
+OTA_UPGRADE_MAX_HW_VERSION_LENGTH         = 2
+'''
```
Comment (wbober) [inline scripts/bootloader/zb_ota/ota_file.py:24]: Background dfu trigger should not be needed in NCS DFU.

File: scripts/bootloader/zb_ota/package.py [hunk 25]
```diff
@@ -0,0 +1,410 @@
+#!/usr/bin/env python3
```
Comment (wbober) [inline scripts/bootloader/zb_ota/package.py:1]: This should probably not be needed here.

File: scripts/bootloader/zb_add_ota_header.py [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+    if isinstance(zb_ota_min_hw_version, int) != isinstance(zb_ota_max_hw_version, int):
+        print('Warning: min/max zigbee ota hardware version is missing. Discarding min/max hardware version.')
+    elif isinstance(zb_ota_min_hw_version, int):
+        if zb_ota_min_hw_version > zb_ota_max_hw_version:
+            print('Warning: zigbee-ota-min-hw-version is higher than zigbee-ota-max-hw-version.')
+
+
+    zigbee_ota_file = OTA_file(app_version_int,
+                                os.path.getsize(app_file),
+                                bytes(open(app_file, 'rb').read()),
```
Comment (tomchy) [inline scripts/bootloader/zb_add_ota_header.py:190]: I'd read the file before constructing the OTA file object, with the appropriate error-handling.

File: scripts/bootloader/zb_add_ota_header.py [hunk 27]
```diff
@@ -0,0 +1,204 @@
...
+    return parser.parse_args()
+
+
+def main():
+
+    args = parse_args()
+
+    app_file = args.application
+    app_version_int = convert_version_string_to_int(args.application_version_string)
+    zb_manufacturer_id = args.zigbee_manufacturer_id if args.zigbee_manufacturer_id is not None else OTA_HEADER_MANUFACTURER_WILDCARD
```
Comment (tomchy) [inline scripts/bootloader/zb_add_ota_header.py:157]: As far as I remember - there is a way to provide default values to optional arguments, so you do not need this assignment.

File: scripts/bootloader/zb_add_ota_header.py [hunk 28]
```diff
@@ -0,0 +1,204 @@
...
+
+def convert_version_string_to_int(s):
+    """Convert from semver string "1.2.3", to integer 10203"""
+    numbers = s.split(".")
+    if len(numbers) != 3:
+        raise print("application-version-string parameter must be on the format x.y.z")
+    js = [10000, 100, 1]
+    return sum([js[i] * int(numbers[i]) for i in range(3)])
+
+def auto_int(x):
```
Comment (tomchy) [inline scripts/bootloader/zb_add_ota_header.py:126]: What is the purpose of this function?

File: scripts/bootloader/zb_add_ota_header.py [hunk 29]
```diff
@@ -0,0 +1,204 @@
...
+
+    parser.add_argument("--application", required=True,
+                        help="The application firmware file.")
+    parser.add_argument("--application-version-string", required=True,
+                        help="The application version string, e.g. '2.7.31'. Will be converted to an integer, e.g. 207031.")
+    parser.add_argument("--zigbee-manufacturer-id", required=False, type=auto_int,
+                        help="Manufacturer ID to be used in Zigbee OTA header.")
+    parser.add_argument("--zigbee-image-type", required=False, type=auto_int,
+                        help="Image type to be used in Zigbee OTA header.")
+    parser.add_argument("--zigbee-comment", required=False, nargs='?',
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_add_ota_header.py:142]: You can specify default value with `default=""`

File: scripts/bootloader/zb_add_ota_header.py [hunk 30]
```diff
@@ -0,0 +1,204 @@
...
+def main():
+
+    args = parse_args()
+
+    app_file = args.application
+    app_version_int = convert_version_string_to_int(args.application_version_string)
+    zb_manufacturer_id = args.zigbee_manufacturer_id if args.zigbee_manufacturer_id is not None else OTA_HEADER_MANUFACTURER_WILDCARD
+    zb_image_type = args.zigbee_image_type if args.zigbee_image_type is not None else OTA_HEADER_IMAGE_TYPE_WILDCARD
+
+    if args.zigbee_comment is None:
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_add_ota_header.py:160]: not needed if default value provided

File: scripts/bootloader/zb_add_ota_header.py [hunk 31]
```diff
@@ -0,0 +1,204 @@
...
+            self.__additional_fields.extend(fields_values)
+        else:
+            self.__additional_fields.append(fields_values)
+
+def convert_version_string_to_int(s):
+    """Convert from semver string "1.2.3", to integer 10203"""
+    numbers = s.split(".")
+    if len(numbers) != 3:
+        raise print("application-version-string parameter must be on the format x.y.z")
+    js = [10000, 100, 1]
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_add_ota_header.py:123]: Since every part of symver can be up to 1024, I would recommend to change js to `js=[1024*1024, 1024, 0]`. 
File version in OTA header is uint32 type, so generated version should fit.
Btw. ZCL spec contains recommended file version definition. One octect for each:
* Application Release 
* Application Build 
* Stack Release 
* Stack Build

File: scripts/bootloader/zb_add_ota_header.py [hunk 32]
```diff
@@ -0,0 +1,204 @@
...
+OTA_HEADER_IMAGE_TYPE_WILDCARD   = 0xFFFF
+OTA_HEADER_STACK_VER_ZIGBEE_PRO  = 0x0002
+
+OTA_HEADER_FIELD_CONTROL_BIT_MASK_HW_VER = (1 << 2)
+OTA_HEADER_MIN_HW_VERSION_LENGTH         = 2
+OTA_HEADER_MAX_HW_VERSION_LENGTH         = 2
+
+class OTA_file:
+
+    def __init__(self,
```
Comment (jangalda-nsc) [inline scripts/bootloader/zb_add_ota_header.py:27]: I would recommend to add typing to all classes and methods

File: scripts/bootloader/zb_add_ota_header.py (hunk 28 above)
Comment (mariuszpos) [reply scripts/bootloader/zb_add_ota_header.py:126]: Converting from hex to int. This conversion is required for argparse int type.

File: scripts/bootloader/zb_add_ota_header.py [hunk 33]
```diff
@@ -0,0 +1,204 @@
...
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+
+import argparse
+import os
+import sys
+import struct
+sys.path.append(os.getcwd())
```
[the remark this answers is not in the document]
Comment (mariuszpos) [reply scripts/bootloader/zb_add_ota_header.py:11]: No

File: scripts/bootloader/zb_add_ota_header.py [hunk 34]
```diff
@@ -0,0 +1,204 @@
...
+
+    # Warn user if minimal/maximum zigbee ota hardware version are not correct:
+    #   * only one of them is given
+    #   * minimum version is higher than maximum version
+    #   * hw_version is inside the range specified by minimum and maximum hardware version
+    if isinstance(zb_ota_min_hw_version, int) != isinstance(zb_ota_max_hw_version, int):
+        print('Warning: min/max zigbee ota hardware version is missing. Discarding min/max hardware version.')
+    elif isinstance(zb_ota_min_hw_version, int):
+        if zb_ota_min_hw_version > zb_ota_max_hw_version:
+            print('Warning: zigbee-ota-min-hw-version is higher than zigbee-ota-max-hw-version.')
```
[the remark this answers is not in the document]
Comment (mariuszpos) [reply scripts/bootloader/zb_add_ota_header.py:185]: It could raise error.

File: scripts/bootloader/zb_add_ota_header.py [hunk 35]
```diff
@@ -0,0 +1,204 @@
...
+    def header(self):
+        return struct.pack(self.__pack_format, *self.__pack_args)
+
+    def __add_optional_fields(self, fields_length, field_control_bit_mask, fields_formatting, fields_values):
+        self.__total_size    += fields_length
+        self.__header_length += fields_length
+        self.__field_control += field_control_bit_mask
+        self.__pack_format   += fields_formatting
+        if type(fields_values) in (tuple, list):
+            self.__additional_fields.extend(fields_values)
```
[the remark this answers is not in the document]
Comment (tomchy) [reply scripts/bootloader/zb_add_ota_header.py:112]: I see the dependency chain. It is a bit complex to untangle this. i think we may leave it as-is.

File: scripts/bootloader/zb_add_ota_header.py (hunk 28 above)
Comment (tomchy) [reply scripts/bootloader/zb_add_ota_header.py:124]: Just please add a comment, explaining the input format.

File: cmake/mcuboot.cmake (hunk 13 above)
Comment (mariuszpos) [reply cmake/mcuboot.cmake:127]: Done

File: subsys/zigbee/Kconfig [hunk 36]
```diff
@@ -208,6 +208,63 @@ config RADIO_STATISTICS
...
+
+config ZIGBEE_OTA_COMMENT
+	string "OTA header string"
+	default ""
+	help
+	  Firmware comment to be used in Zigbee OTA header.
+
+config ENABLE_ZIGBEE_OTA_MIN_HW_VERSION
+	bool "Enable Zigbee OTA minimum hw version"
+	default n
```
Comment (rlubos) [inline subsys/zigbee/Kconfig:247]: No need for `default n`

File: subsys/zigbee/Kconfig [hunk 37]
```diff
@@ -208,6 +208,63 @@ config RADIO_STATISTICS
...
+
+config ZIGBEE_OTA_MIN_HW_VERSION
+	hex "Zigbee OTA minimum hw version"
+	default 0
+	range 0x00 0xFF
+	depends on ENABLE_ZIGBEE_OTA_MIN_HW_VERSION
+
+config ENABLE_ZIGBEE_OTA_MAX_HW_VERSION
+	bool "Enable Zigbee OTA maximum hw version"
+	default n
```
Comment (rlubos) [inline subsys/zigbee/Kconfig:257]: Same here
